subsys/net/ip/tcp2.c

```diff
@@ -1275,6 +1279,110 @@ static enum net_verdict tcp_recv(struct net_conn *net_conn,
 	return NET_DROP;
 }
 
+static uint32_t seq_scale(uint32_t *seq)
+{
+	return *seq + (k_ticks_to_ns_floor32(k_uptime_ticks()) >> 6);
+}
+
+static uint8_t secret_key[16]; /* MD5 128 bits as described in RFC6528 */
+
+static uint32_t tcpv6_init_isn(struct in6_addr *saddr,
+			       struct in6_addr *daddr,
+			       uint16_t sport,
+			       uint16_t dport)
+{
+	struct {
+		uint8_t secret_key[sizeof(secret_key)];
+		struct in6_addr saddr;
+		struct in6_addr daddr;
+		uint16_t sport;
+		uint16_t dport;
+	} buf = {
+		.saddr = *(struct in6_addr *)saddr,
+		.daddr = *(struct in6_addr *)daddr,
+		.sport = sport,
+		.dport = dport
+	};
+
+	uint8_t hash[16];
+	static bool once;
+
+	if (!once) {
+		if (IS_ENABLED(CONFIG_CSPRING_ENABLED)) {
+			sys_csrand_get(secret_key, sizeof(secret_key));
+		} else {
+			sys_rand_get(secret_key, sizeof(secret_key));
+		}
+
+		once = true;
+	}
+
+	memcpy(buf.secret_key, secret_key, sizeof(secret_key));
+	mbedtls_md5_ret((const unsigned char *)&buf, sizeof(buf), hash);
+
+	return seq_scale((uint32_t *)&hash[0]);
```

Comment:
This is potential alignment fault on some archs. `hash` should have `uint32_t` elements to ensure proper alignment (or have explicit alignment annotation, can't remember us using them).